actionmailer/lib/action_mailer/railtie.rb

```diff
@@ -10,6 +10,11 @@ class Railtie < Rails::Railtie # :nodoc:
     config.action_mailer = ActiveSupport::OrderedOptions.new
     config.eager_load_namespaces << ActionMailer
 
+    config.before_eager_load do
+      require "mail"
+      Mail.eager_autoload!
```

Comment:
Can we define a `eager_load!` in `ActionMailer` and move this code to there?

Reply:
`ActionMailer` already has an implicit `eager_load!` as part of `ActiveSupport::Dependencies` which autoloads the `ActionMailer` namespace's autoloads. There doesn't seem to be a clean way to integrate non-ActiveSupport::Dependencies loads into the ActiveSupport::Dependencies eager load gear.

I'd also argue that this is orthogonal to ActionMailer's ActiveSupport::Dependencies load. This change is instead making sure that a third party dependency that ActionMailer introduces also respects and integrates with the Rails application lifecycle. imho, having this before eager load block is the cleanest solution.

Reply:
Yes, I get that. And the place to make sure a third party dependency that Action Mailer introduces is eager loaded is on the Action Mailer constant, since applications what are not Rails applications but use Action Mailer will not execute the railtie.

We should do what we do in [ActiveRecord](https://github.com/rails/rails/blob/e10fd57e98c87774a1dc38761a7d68edee638fdf/activerecord/lib/active_record.rb#L170-L177), defining a `eager_load!` method in `ActionMailer` and call `Mail.eager_autoload!` there. This way we avoid this `before_eager_load` lifecycle and applications that uses Action Mailer but not are rails can also eager load it.

Reply:
Sorry, I didn’t see that prior art, and misinterpreted your comment. I’ll update. Thanks!

Reply:
I'm also sorry for the tone of my original reply here, it's really brusque, patronising and argumentative on reading. I appreciate your time and patience!

Reply:
I've pushed another commit which moves the require and `Mail.eager_autoload!` into `ActionMailer.eager_load!`, like `ActiveRecord.eager_load!` I also tried removing the require, but `Mail` doesn't always seem to be loaded, even after eager loading ActionMailer itself.

Would you like me to squash the commits? They've become pretty noisy... Or will you do that on merge? Or not at all?

Thanks @rafaelfranca!

Reply:
No worries! Your comments were fine to me, it was a valid questioning. The patch looks good, I'll squash with the merge.

Thanks.